File: go-sdk/frame/tape.go
package frame

import (
	"fmt"

	"github.com/ifx-run/ifx/go-sdk/constants"
)
// ValueTypeSize returns payload byte width for a ValueType wire tag.
func ValueTypeSize(tag uint8) (int, error) {
	switch tag {
	case constants.ValueTypeBool, constants.ValueTypeU8, constants.ValueTypeI8:
		return 1, nil
	case constants.ValueTypeU16, constants.ValueTypeI16:
		return 2, nil
	case constants.ValueTypeU32, constants.ValueTypeI32, constants.ValueTypeF32:
		return 4, nil
	case constants.ValueTypeU64, constants.ValueTypeI64, constants.ValueTypeF64:
		return 8, nil
	case constants.ValueTypeU128, constants.ValueTypeI128:
		return 16, nil
	case constants.ValueTypePubkey:
		return 32, nil
	default:
		return 0, fmt.Errorf("unknown ValueType tag %d", tag)
	}
}

// PlanRecordOffsets returns tyOffset, payloadOffset, and endCursor for the next tape append.
// Packed layout: [ty:1][payload:size] with no padding (matches on-chain plan_record_offsets).
func PlanRecordOffsets(cursor uint32, valueTypeTag uint8) (tyOffset, payloadOffset, endCursor uint32, err error) {
	size, err := ValueTypeSize(valueTypeTag)
	if err != nil {
		return 0, 0, 0, err
	}
	tyOffset = cursor
	payloadOffset = cursor + 1
	endCursor = payloadOffset + uint32(size)
	if endCursor > constants.MaxFrameTapeLen {
		return 0, 0, 0, fmt.Errorf("binding does not fit in frame tape (max %d bytes)", constants.MaxFrameTapeLen)
	}
	return tyOffset, payloadOffset, endCursor, nil
}

// RecordByteLength returns bytes for one binding record [ty:1][payload].
func RecordByteLength(valueTypeTag uint8) (int, error) {
	size, err := ValueTypeSize(valueTypeTag)
	if err != nil {
		return 0, err
	}
	return 1 + size, nil
}

File: crates/ifx-core/src/layout/tape.rs
//! Frame tape record layout planning.

use super::error::LayoutError;
use crate::wire::ValueType;

/// Plan the next tape record layout: `(ty_offset, payload_offset)`.
///
/// **Packed layout (no alignment padding):** each record is `[ty:1][payload:ty.size()]`
/// placed immediately after the prior record — `ty @ cursor`, `payload @ cursor + 1`.
/// Off-chain simulators (SDK `planRecordOffsets`) must use identical rules.
pub fn plan_record_offsets(cursor: u32, ty: ValueType) -> Result<(u32, u32), LayoutError> {
    let ty_offset = cursor;
    let payload_offset = cursor
        .checked_add(1)
        .ok_or(LayoutError::TapeOutOfBounds)?;
    let _end = payload_offset
        .checked_add(ty.size() as u32)
        .ok_or(LayoutError::TapeOutOfBounds)?;
    Ok((ty_offset, payload_offset))
}

/// Byte length of one binding record: `[ty:1][payload]`.
pub fn record_byte_length(ty: ValueType) -> u32 {
    1 + ty.size() as u32
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::constants::MAX_FRAME_TAPE_LEN;

    #[test]
    fn packed_u64_chain() {
        let mut cursor = 0u32;
        let (ty0, pay0) = plan_record_offsets(cursor, ValueType::U64).unwrap();
        assert_eq!((ty0, pay0), (0, 1));
        cursor = pay0 + 8;
        let (ty1, pay1) = plan_record_offsets(cursor, ValueType::U64).unwrap();
        assert_eq!((ty1, pay1), (9, 10));
        cursor = pay1 + 8;
        assert_eq!(cursor, 18);
    }

    #[test]
    fn record_byte_length_matches_end_cursor() {
        let cursor = 100u32;
        let ty = ValueType::Pubkey;
        let (_, payload) = plan_record_offsets(cursor, ty).unwrap();
        assert_eq!(payload + ty.size() as u32, cursor + record_byte_length(ty));
    }

    #[test]
    fn tape_len_bounds_constant() {
        assert!(MAX_FRAME_TAPE_LEN <= u16::MAX as u32);
    }
}
